go-controller/pkg/cni/cni.go

```diff
@@ -401,34 +438,46 @@ func getCNIResult(pr *PodRequest, getter PodInfoGetter, podInterfaceInfo *PodInt
 	}, nil
 }
 
+// buildPrimaryUDNPodRequest returns PodRequest for primary UDN interface,
+// it reuturs nil if primary UDN is not requested on the Pod
 func (pr *PodRequest) buildPrimaryUDNPodRequest(
-	pod *corev1.Pod,
 	primaryUDN *udn.UserDefinedPrimaryNetwork,
 ) *PodRequest {
+	if !primaryUDN.Found() {
+		// if primary UDN interface is not needed, return nil
+		return nil
+	}
+	deviceID, deviceInfo, isVFIO := primaryUDN.NetworkDeviceInfo()
 	req := &PodRequest{
 		Command:      pr.Command,
-		PodNamespace: pod.Namespace,
-		PodName:      pod.Name,
-		PodUID:       string(pod.UID),
+		PodNamespace: pr.PodNamespace,
+		PodName:      pr.PodName,
+		PodUID:       pr.PodUID,
 		SandboxID:    pr.SandboxID,
 		Netns:        pr.Netns,
 		IfName:       primaryUDN.InterfaceName(),
 		CNIConf: &ovncnitypes.NetConf{
 			// primary UDN MTU will be taken from config.Default.MTU
 			// if not specified at the NAD
-			MTU: primaryUDN.MTU(),
+			MTU:      primaryUDN.MTU(),
+			DeviceID: deviceID,
 		},
 		timestamp:  time.Now(),
-		IsVFIO:     pr.IsVFIO,
+		IsVFIO:     isVFIO,
 		netName:    primaryUDN.NetworkName(),
 		nadName:    primaryUDN.NADName(),
-		deviceInfo: v1.DeviceInfo{},
+		deviceInfo: *deviceInfo,
 	}
-	req.ctx, req.cancel = context.WithTimeout(context.Background(), 2*time.Minute)
+
+	req.ctx, req.cancel = context.WithCancel(pr.ctx)
 	return req
 }
 
 func (pr *PodRequest) buildPodInterfaceInfo(annotations map[string]string, podAnnotation *util.PodAnnotation, netDevice string) (*PodInterfaceInfo, error) {
+	// pr could be nil if this is PodRequest for primary UDN but primary UDN interface is not needed
+	if pr == nil {
+		return nil, nil
+	}
```

Comment:
i think it would be better to remove this and make it explicit int he callers:

```
primaryUDNPodRequest := pr.buildPrimaryUDNPodRequest(primaryUDN); primaryUDNPodRequest
if primaryUDNPodRequest != nil {
	  primaryUDNPodInfo, err := primaryUDNPodRequest.buildPodInterfaceInfo(annotations, primaryUDN.Annotation(), primaryUDN.NetworkDevice())
	  if err != nil {
		return nil, err
	}
	  klog.Infof("Pod %s/%s primaryUDN podRequest %v podInfo %v", namespace, podName, primaryUDNPodRequest, primaryUDNPodInfo)
}
```

this is what we do in other places in the same function